### PR TITLE
implement `events.on`

### LINKED ISSUE #1750: Add test for async iterable in readline
@ThatOneBro 

image

```js
import * as readline from "node:readline";
import { stdin as input, stdout as output } from "node:process";

const rl = readline.createInterface({ input, output });
for await (const line of rl) {
  console.log(`Received: ${line}`);
}

```

### LINKED ISSUE #2679: `events.on` is not implemented (async iterator for EventEmitter)
The PR for `node:events` (#2604) left this unimplemented. Defined in `events.exports.js`, [reference here](https://github.com/nodejs/node/blob/17c9aa2f7635eccbb132cce20cd45c4cf14fd7be/lib/events.js#L1042), [docs here](https://nodejs.org/api/events.html#eventsonemitter-eventname-options)

### PR DESCRIPTION
**Pull request by nektro**
Closes #2679
Closes #1750

### CODE REVIEW

File: src/js/builtins/ProcessObjectInternals.ts [hunk 1]
```diff
@@ -258,130 +258,8 @@ export function initializeNextTickQueue(process, nextTickQueue, drainMicrotasksF
...
-            this.tail = tail.next;
-            tail.next = null;
-          }
-          return next;
-        }
-      }
-
-      return new FixedQueue();
-    })();
+    const { FixedQueue } = require("$fixed_queue");
```
Comment (paperclover) [inline src/js/builtins/ProcessObjectInternals.ts]:
```suggestion
    const { FixedQueue } = require("internal/fixed_queue");
```

File: src/codegen/internal-module-registry-scanner.ts [hunk 2]
```diff
@@ -57,6 +57,7 @@ export function createInternalModuleRegistry(basedir: string) {
   const requireTransformer = (specifier: string, from: string) => {
     // this one is deprecated
     if (specifier === "$shared") specifier = "./internal/shared.ts";
+    if (specifier === "$fixed_queue") specifier = "./internal/fixed_queue.ts";
```
Comment (paperclover) [inline src/codegen/internal-module-registry-scanner.ts]:
```suggestion

```

File: src/js/node/events.js [hunk 3]
```diff
@@ -354,9 +357,55 @@ function once(emitter, type, options) {
...
+  for (const evName of options?.close || []) {
+    emitter.on(evName, () => {
+      emitter.emit(event, undefined);
+      done = true;
+    });
+  }
+
+  while (!done) {
+    if (!unconsumedEvents.isEmpty()) {
+      yield new Promise(function (resolve, reject) {
```
Comment (Jarred-Sumner) [inline src/js/node/events.js]: This code can be simplified.

```suggestion
      yield Promise.$resolve(unconsumedEvents.shift());
```

File: src/js/node/events.js [hunk 4]
```diff
@@ -354,9 +357,55 @@ function once(emitter, type, options) {
...
+        resolve(unconsumedEvents.shift());
+      });
+    }
+    if (!unconsumedErrors.isEmpty()) {
+      yield new Promise(function (resolve, reject) {
```
Comment (Jarred-Sumner) [inline src/js/node/events.js]:
```suggestion
      yield Promise.$reject(unconsumedErrors.shift())
```

File: src/js/node/events.js [hunk 5]
```diff
@@ -354,9 +357,55 @@ function once(emitter, type, options) {
...
+        reject(unconsumedErrors.shift());
+      });
+    }
+    yield new Promise(function (resolve, reject) {
```
Comment (Jarred-Sumner) [inline src/js/node/events.js]:
```suggestion
    const {promise, reject, resolve} = $newPromiseCapability(Promise);
    unconsumedPromises.push({reject, resolve});
    yield promise;
```

File: src/js/node/events.js [hunk 6]
```diff
@@ -1,8 +1,11 @@
 // Reimplementation of https://nodejs.org/api/events.html
+
 // Reference: https://github.com/nodejs/node/blob/main/lib/events.js
 const { throwNotImplemented } = require("$shared");
+const { FixedQueue } = require("$fixed_queue");
```
Comment (paperclover) [inline src/js/node/events.js]: Can we lazy load this?
```suggestion
let lazyFixedQueue;
```
later:
```ts
const FixedQueue = (lazyFixedQueue ??= require("internal/fixed_queue"))
```
(also, the $shared notation is very old and i was too lazy to migrate all uses to the new resolver. if that import doesnt build please nag me, i dont want to add more `$` prefixed imports)

**Review by paperclover — CHANGES_REQUESTED** [review]

File: test/js/node/events/events-cjs.test.js [hunk 7]
```diff
@@ -1,4 +1,90 @@
...
+
+  const interfaced = createInterface(createReadStream("js/node/child_process/fixtures/child-process-echo-options.js"));
+  const output = [];
+
+  try {
+    for await (const line of interfaced) {
+      output.push(line);
+    }
+  } catch (e) {
+    expect(output).toBe(["// TODO - bun has no `send` method in the process", "process?.send({ env: process.env });"]);
```
Comment (paperclover) [inline test/js/node/events/events-cjs.test.js:91]: It actually does have send

But as per node, it is `undefined` if IPC isn't enabled:
image


edit: oh tihs is just what someone else wrote in a file. lol

**Pull request merged**